Scripting: Per-context script cache, default off

stu-elastic opened the pull request:
* Adds per context settings:
  `script.context.${CONTEXT}.cache_max_size` ~
  `script.cache.max_size`

  `script.context.${CONTEXT}.cache_expire` ~
  `script.cache.expire`

  `script.context.${CONTEXT}.max_compilations_rate` ~
  `script.max_compilations_rate`

* Context cache is used if:
  `script.max_compilations_rate=use-context`.  This
  value is dynamically updatable, so users can
  switch back to the general cache if desired.

* Settings for context caches take the first value 
  that applies:
  1) Context specific settings if set, eg
     `script.context.ingest.cache_max_size`
  2) Correlated general setting is set to the non-default 
     value, eg `script.cache.max_size`
  3) Context default

The reason for 2's inclusion is to allow an easy
transition for users who've customized their general
cache settings.

Using the general cache settings for the context caches
results in higher effective settings, since they are 
multiplied across the number of contexts.  So a general
cache max size of 200 will become 200 * # of contexts.
However, this behavior it will avoid users snapping to a
value that is too low for them.


Refs: #50152

stu-elastic committed:
Scripting: Per-context script cache, default off

* Adds per context settings:
  `script.context.${CONTEXT}.cache_max_size` ~
    `script.cache.max_size`

  `script.context.${CONTEXT}.cache_expire` ~
    `script.cache.expire`

  `script.context.${CONTEXT}.max_compilations_rate` ~
    `script.max_compilations_rate`

* General script cache is used if there are no cache settings or if
  the general settings are used.

* Context script cache is used if there are any context cache settings
  and no general cache settings.

* Mixed context and general cache settings are an error.

* Context cache settings are dynamically updatable.  If the general
  cache is used, these updates are rejected.

* Existing general cache setting, `script.max_compilations_rate` remains
  dynamically updatable.  Updates are rejected if context caches are in
  use.

jdconrad commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -219,7 +383,29 @@ boolean compilationLimitsEnabled() {
+                        // Here we try to detect if it's a spurious update to the default.
+                        // This _will_ allow through bad updates to the default.
+                        if (SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.getDefault(Settings.EMPTY).equals(s) &&
+                            SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.exists(settings) == false) {
```
After looking through the code more carefully, we can't do this because settings is never actually updated for dynamic settings. This would only determine if this setting exists at start-up. So if a user were to dynamically update this setting we would be using old information.

stu-elastic replied:
What do you think about changing this to:

`SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.getDefault(Settings.EMPTY).equals(s)`

That will detect non-changes.

rjernst replied:
That would only detect if the value was different than the default. The user could explicitly set the default through a dynamic update?

stu-elastic replied:
Yes, the user can set that and we will not reject it.  I address this in the comment on [line 397](https://github.com/elastic/elasticsearch/pull/52855/files#diff-e7f789c9bbf345278cfb259a0b9ac87fR397):
```
// This validator is run on every settings update.
// Here we try to detect if it's a spurious update to the default.
// This _will_ allow through bad updates to the default.
```

jdconrad replied:
@stu-elastic Thanks for walking me through this. I believe you are correct that if we remove the exists check, this will behave as expected.

stu-elastic replied:
Now we have a special `use-context` value.

jdconrad reviewed (changes requested):
I spent more time thinking about how we use `SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED` as a sentinel value for determining whether we have overlapping deprecated settings versus new context settings, and I think we need to come up with something different :/ See my comment at those specific lines.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Scripting)

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -93,15 +94,52 @@
             };
 
-    public static final Setting<Integer> SCRIPT_CACHE_SIZE_SETTING =
+    public static final Setting<Integer> SCRIPT_CACHE_SIZE_SETTING_DEPRECATED =
```
I don't think we've used this naming pattern of suffixing with _DEPRECATED before. If we are going to change the object name, perhaps instead we could refer to this as the "general" version since that is the terminology you used elsewhere? eg SCRIPT_GENERAL_CACHE_SIZE_SETTING. We also need the Property.Deprecated on these.

stu-elastic replied:
> We also need the Property.Deprecated on these.

That will be added in a future change, see [step 5](https://github.com/elastic/elasticsearch/issues/50152#issuecomment-587927041)

stu-elastic replied:
Changed to `SCRIPT_GENERAL_`, leaving the deprecated property addition to a future change.

stu-elastic replied:
As mentioned, deprecation will happen in a future change.  There's a bunch of tests, etc that need to change in tandem with that.

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -93,15 +94,52 @@
             new Setting<>("script.max_compilations_rate", "75/5m", MAX_COMPILATION_RATE_FUNCTION, Property.Dynamic, Property.NodeScope);
 
+    // Per-context settings
+    static final String contextPrefix = "script.context.";
```
call caps case?

stu-elastic replied:
Changed.

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -201,13 +240,138 @@ public ScriptService(Settings settings, Map<String, ScriptEngine> engines, Map<S
+        Set<String> generalKeys = new HashSet<>();
+        Set<String> contextKeys = new HashSet<>();
+        for (String key: settings.keySet()) {
+            if (SCRIPT_CACHE_SIZE_SETTING_DEPRECATED.match(key) ||
```
Why are we not just using the exists method?

stu-elastic replied:
I can pull these out.  I figured I'd just keep the same shape since it doesn't work for Affix settings.
```java
ScriptService.SCRIPT_CACHE_EXPIRE_SETTING.exists(
            Settings.builder()
                .put("script.context.baz.cache_expire", TimeValue.timeValueMillis(1000))
                .build()
        )
```
is false.

stu-elastic replied:
Using exists now.

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -201,13 +240,138 @@ public ScriptService(Settings settings, Map<String, ScriptEngine> engines, Map<S
+                generalKeys.add(key);
+
+            } else if (
+                SCRIPT_CACHE_SIZE_SETTING.getRawKey().match(key) ||
```
This looks like the case for cross setting validation. Whether the general cache should be created should strictly be based on any general cache settings existing. Validating both sets of settings don't exist at the same time is a separate issue, and can be done by the settings validators themselves.

stu-elastic replied:
> Validating both sets of settings don't exist at the same time is a separate issue, and can be done by the settings validators themselves.

Let's discuss how this can be done.

stu-elastic replied:
We're no longer worrying about cross validation.  Now we allow any combo of settings to be set and flip between context and general caches using `script.max_compilations_rate=use-context`.

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -219,7 +383,29 @@ boolean compilationLimitsEnabled() {
+                        generalCache.setMaxCompilationRate(r);
+                    }
+                },
+                s -> {
```
I think we need to consider changing the signature here to allow the old value to be passed through. We have it in the update consumer within Setting. Also, using the cross setting validation should allow simpler validation here I think. See the other validate methods that can be implemented on Setting.Validator.

stu-elastic replied:
We've taken a step back and are changing the semantics here as mentioned above.

rjernst reviewed (commented):
I have a few comments

stu-elastic committed:
Setting SCRIPT_*_DEPRECATED -> SCRIPT_GENERAL_*; contextPrefix -> CONTEXT_PREFIX, use `exists` for non-affix settings; Remove general compilation rate exists check in dynamic validator

stu-elastic commented:
Blocking this change on fixing affix settings in https://github.com/elastic/elasticsearch/issues/52933

stu-elastic committed:
Settings: AffixSettings as validator dependencies

Allow AffixSetting as validator dependencies.  If a validator
specifies AffixSettings as a dependency, then `validate(T, Map)`
will have the concrete setting in a map.

Fixes: #52933

stu-elastic committed:
Use validators for TimeSettings, use atomic ref to change context settings

stu-elastic committed:
Get affix match any

stu-elastic committed:
WIP flag

stu-elastic committed:
Add script.context.cache_enabled setting, use single cache updater to process updates

stu-elastic committed:
Allow "use-context" setting to flip back and forth between context and general caches

stu-elastic committed:
Merge branch 'master' into fix/50152-painless-limit-per-context__02__context_specific

stu-elastic committed:
Remove commented out tests, unnecessary settings updates, simpile cache checking

stu-elastic committed:
Add some cache object tests

stu-elastic committed:
Test per context nodes

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into fix/50152-painless-limit-per-context__02__context_specific

jdconrad commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -468,11 +515,259 @@ public ScriptLanguagesInfo getScriptLanguages() {
+     */
+    static class CacheUpdater {
+        private final Object lock = new Object();
+        private static final Tuple<Integer, TimeValue> ZERO = new Tuple<>(0, TimeValue.ZERO);
```
`ZERO` is pretty vague as a name here, but I see it matches TimeValue. Maybe this could be `SCRIPT_COMPILATION_RATE_ZERO` or something like that to match `SCRIPT_COMPILATION_RATE`.

stu-elastic replied:
Changed.

jdconrad reviewed (approved):
LGTM. Thanks for all the work here. One very minor comment.

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into fix/50152-painless-limit-per-context__02__context_specific

stu-elastic committed:
ZERO -> SCRIPT_COMPILATION_RATE_ZERO

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -219,7 +252,16 @@ boolean compilationLimitsEnabled() {
     void registerClusterSettingsListeners(ClusterSettings clusterSettings) {
         clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_SIZE_IN_BYTES, this::setMaxSizeInBytes);
-        clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_COMPILATIONS_RATE, compiler::setMaxCompilationRate);
+        clusterSettings.addSettingsUpdateConsumer(SCRIPT_GENERAL_MAX_COMPILATIONS_RATE, cacheUpdater::setMaxCompilationRate);
```
One problem that seems to make this more complicated is having individual update consumers. There is an existing method we can use that takes a list of settings, and passes all the values when any are updated. I think using this pattern will make the code simpler (less locking and special cases). The pattern I'm imagining is something like an update consumer for the following scenarios:

- general settings -> this changes the general cache, and handles flipping between general and per context
- per context + general settings -> the general settings along with concrete settings for each context. if the general cache is used, this would result in an error if any of the per context settings exist. otherwise the ScriptCache for that context would be looked up. All 3 setting values should then be passed into a single update method on ScriptCache, so it can be rebuilt once whether one or all 3 settings change (or the fallbacks for size/expire changed).

stu-elastic replied:
Done.

rjernst reviewed (changes requested):
This is looking good. I like the new semantics. However I think we can further simplify.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -93,14 +102,45 @@
+
+    // script.context.<context-name>.{cache_max_size, cache_expire, max_compilations_rate}
+
+    // TODO(stu): set to sentinel value when we start using per-context defaults
```
Future looking comment based on implementation.  Will remove.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -219,7 +252,16 @@ boolean compilationLimitsEnabled() {
+        clusterSettings.addAffixUpdateConsumer(SCRIPT_CACHE_SIZE_SETTING,
+                                               cacheUpdater::setScriptCacheSize,
+                                               cacheUpdater::contextExists);
+        clusterSettings.addAffixUpdateConsumer(SCRIPT_CACHE_EXPIRE_SETTING,
```
They were never dynamically updatable before, why add them for deprecated settings?

stu-elastic replied:
From direct convo, not necessary to add.

stu-elastic committed:
SCRIPT_MAX_COMPILATIONS_RATE += _SETTING, CacheUpdater:contextExists -> ScriptService, throw to assert

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -93,14 +102,45 @@
         Setting.intSetting("script.max_size_in_bytes", 65535, 0, Property.Dynamic, Property.NodeScope);
-    public static final Setting<Tuple<Integer, TimeValue>> SCRIPT_MAX_COMPILATIONS_RATE =
-            new Setting<>("script.max_compilations_rate", "75/5m", MAX_COMPILATION_RATE_FUNCTION, Property.Dynamic, Property.NodeScope);
+    public static final Setting<Tuple<Integer, TimeValue>> SCRIPT_GENERAL_MAX_COMPILATIONS_RATE =
```
Appended.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -93,14 +102,45 @@
+            key -> Setting.positiveTimeSetting(key, TimeValue.timeValueMillis(0), Property.NodeScope, Property.Dynamic));
+
+    // TODO(stu): set to sentinel value when we start using per-context defaults
+    public static final Setting.AffixSetting<Tuple<Integer, TimeValue>> SCRIPT_MAX_COMPILATIONS_RATE =
```
Mimeographed.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -304,7 +346,12 @@ void setMaxSizeInBytes(int newMaxSizeInBytes) {
-        return compiler.compile(context, scriptEngine, id, idOrCode, type, options);
+        ScriptCache scriptCache = this.cache.get().contextCache.get(context.name);
+        if (scriptCache == null) {
+            throw new IllegalArgumentException("script context [" + context.name + "] has no script cache");
```
Asserted.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -468,11 +515,260 @@ public ScriptLanguagesInfo getScriptLanguages() {
+        }
+
+        void contextExists(String context, Object o) {
+            if (contextSettings.containsKey(context) == false) {
```
Moved.

stu-elastic committed:
MAX_COMPILATION_RATE_FUNCTION check of USE_CONTEXT_RATE_KEY moved to lambda in SCRIPT_GENERAL_MAX_COMPILATIONS_RATE_SETTING

stu-elastic committed:
flip assert

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -61,10 +62,18 @@
     // this allows you to easily define rates
     static final Function<String, Tuple<Integer, TimeValue>> MAX_COMPILATION_RATE_FUNCTION =
             (String value) -> {
+                if (value.equals(USE_CONTEXT_RATE_KEY)) {
```
Moved.

stu-elastic committed:
script.context.$CONTEXT.{cache_max_size,cache_expire} fallback to general settings

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -93,14 +102,45 @@
+    public static final Setting.AffixSetting<TimeValue> SCRIPT_CACHE_EXPIRE_SETTING =
+        Setting.affixKeySetting(CONTEXT_PREFIX,
+            "cache_expire",
+            key -> Setting.positiveTimeSetting(key, TimeValue.timeValueMillis(0), Property.NodeScope, Property.Dynamic));
```
Added fallback.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -93,14 +102,45 @@
+    public static final Setting.AffixSetting<Integer> SCRIPT_CACHE_SIZE_SETTING =
+        Setting.affixKeySetting(CONTEXT_PREFIX,
+            "cache_max_size",
+            key -> Setting.intSetting(key, 100, 0, Property.NodeScope, Property.Dynamic));
```
Added fallback.

stu-elastic committed:
use validator in group settings updater, validate all settings together

stu-elastic committed:
Update dependencies on context cache updaters

stu-elastic committed:
Add conflict, fallback, and validator tests

stu-elastic committed:
CacheHolder settings tests

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -468,11 +515,260 @@ public ScriptLanguagesInfo getScriptLanguages() {
+
+        final boolean generalExpireSet;
+        final Set<String> contextExpireSet;
+        final TimeValue expireDefault = SCRIPT_CACHE_EXPIRE_SETTING.getDefault(Settings.EMPTY);
```
Using fallback settings, thanks for pointing it out.

stu-elastic commented on server/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -468,11 +515,260 @@ public ScriptLanguagesInfo getScriptLanguages() {
+        ) {
+            this.cacheRef = cacheRef;
+
+            generalSizeSet = SCRIPT_GENERAL_CACHE_SIZE_SETTING.exists(settings);
```
Using fallback settings.

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into fix/50152-painless-limit-per-context__02__context_specific

jdconrad reviewed (approved):
LGTM after all these changes. Thanks for cleaning this up so much.

rjernst reviewed (approved):
LGTM

stu-elastic commented:
master: 070ea7e
7.x:  cdbee32
